libs/compress/src/main/java/org/opensearch/compress/spi/CompressionProvider.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.compress.spi;
+
+import org.opensearch.compress.ZstdCompressor;
+import org.opensearch.core.compress.Compressor;
+import org.opensearch.core.compress.spi.CompressorProvider;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.List;
+
+/**
+ * Additional "optional" compressor implementations provided by the opensearch compress library
+ *
+ * @opensearch.internal
+ */
+public class CompressionProvider implements CompressorProvider {
```

Comment:
Since this class is only used in SPI registration and will likely never be referenced directly in code, I'd go with a more verbose name. Since it is the CompressorProvider provided by the `opensearch-compress` library, how about `OpenSearchCompressCompressorProvider`?

Reply:
Just curious, what's the reason for a verbose name?

Reply:
`CompressionProvider` offers no clue as to what kind of `CompressorProvider` this is. They are so similarly named but yet subtly different that it strikes me as quite confusing if I come across both while searching for classes in my IDE, for example.

Reply:
> For example, `ServerCompressorProvider` is very clear, but `CompressionProvider` kind of looks like a typo to me.

So we use compression in 2 places:
 - Apache Lucene codec 
 - Repostories 

The  `CompressionProvider` (or `CompressorProvider`) to me is better than `ServerCompressorProvider` because there is nothing really server specific there (and I would expect to `ClientXxxx` counterpart to appear somewhere).

Reply:
> The `CompressionProvider` (or `CompressorProvider`) to me is better than `ServerCompressorProvider` because there is nothing really server specific there

That was the reason for keeping it as is. I also think the package name `org.opensearch.core.compress.CompressorProvider` is pretty self explanatory that this is a core provider for Compressor implementations?

Similarly, the Lucene codec compression namespace is also pretty self explanatory (`org.apache.lucene.codecs.compressing.*`). The `DeflateCompressor` `LZ4FastCompressor` and `LZ4HighCompressor` are also private classes in Lucene, so there should be no ambiguity.

Reply:
Also, there's nothing stopping someone from doing something like:

```
public class MyOpenSearchZSTDCompressionForLuceneCodec  extends org.apache.lucene.codecs.compressing.Compressor {

  private static final org.opensearch.compress.ZstdCompressor opensearchZSTDDelegate = CompressorRegistry.getCompressor(ZstdCompressor.NAME);

  @Override
  public void compress(ByteBuffersDataInput buffersInput, DataOutput out) {
     // delegate compression to OpenSearch
     BytesReference br = BytesReference.fromByteArray(...


     BytesReference compressed = opensearchZSTDDelegate.compress(br);
     ....
  }
```

Reply:
My nitpick was pretty narrowly focused on `CompressionProvider implements CompressorProvider`. Why is the word `Compression` substituted for `Compressor` in the class name? I'd honestly prefer `CompressorProvider implements org.opensearch.core.compress.CompressorProvider` unless there's a good answer to that question.

Reply:
> Why is the word `Compression` substituted for `Compressor` in the class name?

You mean the compress library's concrete `CompressionProvider` class? Because that classname is only used in the META-INF file for SPI `ServiceLoader`. Developers will not call that so the name is irrelevant. I can rename it to clash w/ the `CompressorProvider` interface (who's name does matter) and use the implements FQCN. I just personally try to avoid FQCN at all cost.

Reply:
@andrross  Do you want me to change the name of the `:lib:opensearch-compress` SPI loaded class from `CompressionProvider` to the clashing `CompressorProvider` and implement by FQCN? Or are we fine w/ `:libs:opensearch-compress` keeping the `CompressionProvider` concrete classname since it's never actually referenced outside the `resources/META-INF/services/org.opensearch.core.compress.spi.CompressorProvider` file?

Reply:
> Because that classname is only used in the META-INF file for SPI ServiceLoader. Developers will not call that so the name is irrelevant.

Agree this name doesn't matter as developers won't call it. However, my IDE will index it, and if I search for the class name c-o-m-p-r-e-s-s looking for `CompressorProvider` this one will pop up and I have to mentally sort through `CompressionProvider` and `CompressorProvider`. I also dislike working with FQCN in Java hence the reason I suggested the verbose name of the format `<library name>CompressorProvider`. I fully acknowledge this is a nitpick :)

Reply:
When we enable JPMS we won't be exporting this class so the IDE won't display it in any packages other than `o.o.compress`. Another nice property of the modularity. 🙂